sqlite3_opt_serialize.go

```diff
@@ -0,0 +1,68 @@
+// +build !sqlite_omit_deserialize
+
+package sqlite3
+
+/*
+#ifndef USE_LIBSQLITE3
+#include <sqlite3-binding.h>
+#else
+#include <sqlite3.h>
+#endif
+#include <stdlib.h>
+#include <stdint.h>
+*/
+import "C"
+
+import (
+	"fmt"
+	"unsafe"
+)
+
+// Serialize returns a byte slice that is a serialization of the database.
+// If the database fails to serialize, a nil slice will be returned.
+//
+// See https://www.sqlite.org/c3ref/serialize.html
+func (c *SQLiteConn) Serialize(schema string) []byte {
+	if schema == "" {
+		schema = "main"
+	}
+	var zSchema *C.char
+	zSchema = C.CString(schema)
+	defer C.free(unsafe.Pointer(zSchema))
+
+	var sz C.sqlite3_int64
+	ptr := C.sqlite3_serialize(c.db, zSchema, &sz, 0)
+	if ptr == nil {
+		return nil
+	}
+	defer C.sqlite3_free(unsafe.Pointer(ptr))
+	return C.GoBytes(unsafe.Pointer(ptr), C.int(sz))
```

Comment:
Casting `sqlite3_int64` to `int` could overflow on some systems.

Reply:
@rittneje -- since the signature of `GoBytes` requires `C.int` should I copy the data byte-by-byte in a loop?

Reply:
Since Go represents the length (and capacity) of a slice as an `int`, the potential overflow will be an issue on 32-bit systems no matter what. Thus in that case you have no choice but to return an error. However, on a 64-bit system, you should be safe.

Regardless, there is no guarantee about `C.int` itself on either system. It is _likely_ to be 32-bits in both cases, but the C standard has no strict requirement.

I would not copy byte-by-byte as that will be very inefficient, particularly for large databases. In the interest of keeping things simple (at least for your initial implementation), I think we can just do some checking around `C.sizeof_int` like so:
```go
if C.sizeof_int < 64 {
    maxSize := C.sqlite3_int64(1) << C.sizeof_int - 1
    if sz > maxSize {
        return nil, fmt.Errorf("sqlite3: serialized database is too large (%d bytes)", maxSize)
    }
}
return C.GoBytes(unsafe.Pointer(ptr), C.int(sz)), nil
```

If we wanted to support larger databases on 64-bit systems, I think it would have to be like this:
```go
// Note: math.MaxInt was added in Go 1.17. We'd have to do something else for older versions.
if sz > C.sqlite3_int64(math.MaxInt) {
    return nil, fmt.Errorf(...)
}

cBuf := *(*[]byte)(unsafe.Pointer(&reflect.SliceHeader{
    Data: unsafe.Pointer(ptr),
    Len: int(sz),
    Cap: int(sz),
}))

res := make([]byte, int(sz))
copy(res, cBuf)
return res
```

Reply:
Fixed in code.